landscaper/gardenlet/pkg/gardenlet_delete.go

```diff
@@ -0,0 +1,126 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pkg
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	"github.com/gardener/gardener/pkg/utils/retry"
+
+	"github.com/gardener/gardener/landscaper/gardenlet/pkg/applier"
+)
+
+// Delete removes all deployed Gardenlet resources from the Seed cluster.
+func (g Landscaper) Delete(ctx context.Context) error {
+	shootList := &gardencorev1beta1.ShootList{}
+	err := g.gardenClient.Client().List(ctx, shootList)
+	if err != nil {
+		return fmt.Errorf("cannot delete Seed '%s'. Failed to check if Seed is still in use by shoot(s): %v", g.gardenletConfiguration.SeedConfig.Name, err)
+	}
+
+	seed := &gardencorev1beta1.Seed{ObjectMeta: metav1.ObjectMeta{
+		Name: g.gardenletConfiguration.SeedConfig.Name,
+	}}
+
+	exists, err := g.seedExists(ctx, seed)
+	if err != nil {
+		return fmt.Errorf("failed to check if Seed %q exists in the Garden cluster. Aborting. %v", g.gardenletConfiguration.SeedConfig.Name, err)
+	}
+
+	if exists {
+		if isSeedUsedByAnyShoot(g.gardenletConfiguration.SeedConfig.Name, shootList.Items) {
+			return fmt.Errorf("cannot delete seed '%s' which is still used by at least one shoot", g.gardenletConfiguration.SeedConfig.Name)
+		}
+
+		err := g.waitForSeedDeletion(ctx, seed)
+		if err != nil {
+			return err
+		}
+	}
+
+	chartApplier := g.seedClient.ChartApplier()
+	applier := applier.NewGardenletChartApplier(chartApplier, map[string]interface{}{}, GetChartPath())
+	if err := applier.Destroy(ctx); err != nil {
+		return fmt.Errorf("failed to delete the Gardenlet resources from the Seed cluster %q: %v", g.gardenletConfiguration.SeedConfig.Name, err)
+	}
+
+	g.log.Infof("Successfully deleted Gardenlet resources for Seed %q", g.gardenletConfiguration.SeedConfig.Name)
+	return nil
+}
+
+// waitForSeedDeletion waits until the Seed resource has been deleted from the Garden cluster
+func (g Landscaper) waitForSeedDeletion(ctx context.Context, seed *gardencorev1beta1.Seed) error {
+	if err := retry.UntilTimeout(ctx, 10*time.Second, 2*time.Minute, func(ctx context.Context) (done bool, err error) {
+		err = g.gardenClient.Client().Delete(ctx, seed)
+		if err != nil {
+			g.log.Infof("Error setting deletion timestamp on seed %q: %v", g.gardenletConfiguration.SeedConfig.Name, err)
```

Comment:
```suggestion
			g.log.Infof("Error deleting seed %q: %v", g.gardenletConfiguration.SeedConfig.Name, err)
```
I don't think we should talk about deletion timestamps, this is an implementation detail, the actual operation that is failing is "delete"